---
title: "Pseudotime analysis on Macro/Mono subset 1"
author: "Kaspar Bresser"
date: "20/10/2021"
output: 
  github_document:
    toc: true
  html_document: 
    theme: simplex
    highlight: pygments
#    code_folding: show
    self_contained: TRUE
    toc: yes
    toc_float:
      collapsed: no
---



```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      message=FALSE,
                      autodep  = TRUE,
                      cache = FALSE,
                      fig.width = 5,
                      fig.asp = 0.618,
                      fig.align = "center")
```

In the immune compartment of the scRNAseq dataset, we identified 2 Macro/Mono subsets. In subset 1 we could see an enrichment and depletion of certain MetaCells. To dive into this further, we made use of an pseudotime analysis to assess if expression of certain genes was associated with a QPCTL-proficient or deficient state.

Import packages and data needed for the analysis.

```{r loading, echo=FALSE}
library(here)
library(RColorBrewer)
library(Seurat)
library(metacell)
library(slingshot)
library(tradeSeq)
library(tidyverse)
library(SummarizedExperiment)
library(tidyverse)

scdb_init(here("Data", "Metacell_files_imm"))
mc <- scdb_mc("QPCTL_Imm_MC")
MC.graph <- scdb_mc2d("QPCTL_Imm_MC")


seurat.obj <- read_rds(here("Data", "seurat_immune.rds"))
```


## Prepare dataset

Add MetaCells of interest (1-3) as metadata. 

```{r add_MC}
MCs <- mc@mc[mc@mc %in% c(1,2,3)]

seurat.obj <- AddMetaData(object = seurat.obj, metadata = MCs, col.name = "MCs")
```

Add MetaCell coordinates to seurat object as a custom dimensionality reduction. First merge metacell coords with some metadata columns, to make sure the length of the matrix matches that of the seurat object. 

```{r add_coords}
coords <- data.frame( row.names =  names(MC.graph@sc_y),
                  MC_1 = MC.graph@sc_x,
                  MC_2 = MC.graph@sc_y)
head(coords)

coords <- merge(coords, seurat.obj@meta.data[,c(13,14)], by = 0, all = T)

coords %>% 
  select(Row.names, MC_1, MC_2) %>% 
  column_to_rownames("Row.names") %>% 
  as.matrix() -> coords

seurat.obj[["mcs"]] <- CreateDimReducObject(embeddings = coords, key = "MC_", assay = DefaultAssay(seurat.obj))
```

We can now use the MetaCell 2d coordinates as you would any other dimensional reduction in downstream functions. 

Lets subset on the MetaCells of interest. 

```{r subset_data}
Idents(seurat.obj) <- "MCs"

DimPlot(seurat.obj, reduction = "mcs", pt.size = 0.5)


seurat.obj <- subset(seurat.obj, cells = row.names(subset(seurat.obj@meta.data, !is.na(MCs) ) ))


DimPlot(seurat.obj, reduction = "mcs", pt.size = 1.5)
```

## Pseudotime analysis

### Perform slingshot

Normalize data, and perform [slingshot](https://bioconductor.org/packages/release/bioc/html/slingshot.html) analysis to get pseudotime values. 


```{r perform_slingshot}
# add normalized counts
seurat.obj <- NormalizeData(object = seurat.obj, normalization.method = "LogNormalize", scale.factor = 10000)

# create sce and perform slingshot
immune.sce <- as.SingleCellExperiment(seurat.obj)
immune.sce <- immune.sce[,names(MCs)]
immune.sce <- slingshot(data = immune.sce, clusterLabels = "MCs", reducedDim = "MCS")
```


Pseudotime values are now saved in the `SingleCellExperiment` object.


### Test for genes associated with pseudotime

Next we'll get both the log-transformed and untransformed expression values. Focus on the genes that were used for the MetaCell generation. 

```{r get_expression}
# get vargenes used for MC generation
vargenes <- scdb_gset("QPCTL_Imm_feats_filt")
vargenes <- names(vargenes@gene_set)

# get the log transformed counts from the sce object, subset for vargenes and cellcodes
logexpr <- as.matrix(seurat.obj@assays$RNA@data)[vargenes, names(MCs)]
# same, but to get non-transformed data
expr <- as.matrix(seurat.obj@assays$RNA@counts)[vargenes, names(MCs)]
```

Extract the pseudotime values and cell-weights that were generated by slingshot.

```{r get_pseudo}
lineages <- slingPseudotime(immune.sce)
cellWeights <- slingCurveWeights(immune.sce)
```

We'll be testing for genes associated to pseudotime using General Additive Models, we can use `evaluateK` to assess the number of knots that would be optimal for GAM fitting. Note that we use untransformed expression data here. See [here](http://bioconductor.org/packages/release/bioc/vignettes/tradeSeq/inst/doc/fitGAM.html) for more info on evaluateK.


```{r knots, fig.width=8}
# Evaluate the number of knots needed for GAM fitting
icMat2 <- evaluateK(counts = expr, pseudotime = lineages, cellWeights = cellWeights,
                    k=3:7, nGenes = 100, verbose = FALSE, plot = TRUE)

```

We'll fit the GAMs with using `nknots = 5`

```{r fitGAM}
immune.sce <- fitGAM(counts = expr, sds = SlingshotDataSet(immune.sce), nknots = 5)
```

Next, test for dynamic expression, output will have genes associated with pseudotime

```{r test_association}
assoRes <- associationTest(immune.sce)

(assoRes <- as_tibble(assoRes, rownames = "gene"))
```


### Cluster and plot associations

First, get significantly associated genes.

```{r sig_genes}
assoRes %>% 
  mutate(padj = p.adjust(pvalue, method = "bonferroni")) %>% 
  filter(padj < 0.05) %>% 
  # slice_max(order_by = waldStat, n = 30) %>% 
  pull(gene) -> genes

```

Subset log-expression matrix to genes associated with pseudotime and scale.

```{r scale_genes}
hclust_matrix <- logexpr[genes, ]

hclust_matrix %>% 
  # transpose the matrix so genes are as columns
  t() %>% 
  # apply scalling to each column of the matrix (genes)
  scale() %>% 
  # transpose back so genes are as rows again
  t() -> hclust_matrix
```

Perform hierarchical clustering.

```{r hclust,  fig.keep='all',fig.width=8}
gene_dist <- dist(hclust_matrix)
gene_hclust <- hclust(gene_dist, method = "complete")

# The default `plot()` function can be used to produce a simple dendrogram
plot(gene_hclust, labels = FALSE)
abline(h = 31, col = "brown", lwd = 2) # add horizontal line to illustrate cutting dendrogram
```

Decided on 6 clusters, use cuttree to set clusters

```{r add_clusters}
cutree(gene_hclust, k = 6) %>% 
  # turn the named vector into a tibble
  enframe("gene", "cluster") -> gene.cluster

gene.cluster %>% 
  group_by(cluster) %>% 
  nest() %>% 
  pull(data) %>% 
  map( pull, gene)
```


Make dataframe containing pseudotimes per cellcode

```{r get_pseudos}
lineages %>%
  as_tibble(rownames = "cellcode") %>% 
  dplyr::rename( pseudotime = "curve1") %>% 
  drop_na() -> pseudotimes

pseudotimes
```

Make dataframe containing scaled logcounts of genes per cellcode

```{r get_scales_expr}
# define function for sd scaling of count data
scale_this <- function(x){
  (x - mean(x, na.rm=TRUE)) / sd(x, na.rm=TRUE)
}

logexpr[genes,] %>% 
  as_tibble(rownames = "gene") %>% 
  pivot_longer(cols = -gene, names_to = "cellcode", values_to = "logcount") %>% 
  filter(cellcode %in% pseudotimes$cellcode) %>% 
  group_by(gene) %>% 
  mutate(scaled_counts = scale_this(logcount)) -> logcounts

logcounts
```

Make combined table with pseudotimes, GAM-cluster and gene-expression

```{r combine_data}
pseudotimes %>% 
  left_join(logcounts, by = "cellcode") %>% 
  left_join(gene.cluster, by = "gene") -> GAM.data

GAM.data

write_tsv(x = GAM.data, here("Output", "Imm_gene_clusters_pseudotime.tsv"))
```

Plot the gene(GAM)-clusters.

```{r plot_clusters}
ggplot(GAM.data, aes(x = pseudotime, y = scaled_counts))+
  geom_smooth(aes(group =gene), se = F, color = "lightgrey", method = "gam", size = 0.75)+
  geom_smooth(method = "gam")+
  scale_x_reverse()+
  facet_wrap(~cluster, scales = "free_y")

ggsave(filename = here("Figs", "immune", "Pseudotime_clusters.pdf"), width = 7,height = 4, useDingbats = F )
```

Now focus on some individual genes to visualize. 

```{r get_focus_genes, fig.width=4, fig.asp=2}
plot.genes <- c("Eps8", "Creg1", "Pld3","Serpinb6a","Ecm1","Cd5l",
                "Ly6c2", "Ms4a4c","Lyz1", "Lamp1", "Cstb", "Ctsl", 
                "Ccr2", "H2-Ab1", "Cd74", "Mrc1", "Retnla", "Fn1", 
                "Lipa", "Ctsd")

GAM.data %>% 
  filter(gene %in% plot.genes) %>% 
    ggplot(aes(x = pseudotime, y = logcount))+
      geom_point(size = 0.2, color = "grey")+ 
      geom_smooth(method = "gam")+
      scale_x_reverse()+
      facet_wrap(~gene, scales = "free_y", ncol = 4)

ggsave(filename = here("Figs", "immune", "Pseudotime_genes_points.pdf"), width = 4,height = 7, useDingbats = F )
```


And same plots, but without the points. 

```{r focus_genes, fig.width=4, fig.asp=2}
GAM.data %>% 
  filter(gene %in% plot.genes) %>% 
    ggplot(aes(x = pseudotime, y = logcount))+
      geom_smooth(method = "gam")+
      scale_x_reverse()+
      facet_wrap(~gene, scales = "free_y", ncol = 4)

ggsave(filename = here("Figs", "immune", "Pseudotime_genes.pdf"), width = 4,height = 7, useDingbats = F )
```


## Enrichment of QPCTL-prof or -def

Finally, we'll plot the moving sums of QPCTL-proficient and deficient cell counts across pseudotime. Will use the `RcppRoll` package.

Add a count column, and normalize for hashtag size. Also collapse samples by QPCTL proficient and deficient. 

```{r get_hashtags}
library(RcppRoll)

### Plot rolling average cell counts across pseudotime
seurat.obj@meta.data %>% 
  as_tibble(rownames = "cellcode") %>% 
  mutate(hash.ID = case_when(hash.ID == "Doublet" ~ HTO_maxID,
                             TRUE ~ hash.ID)) %>% 
  dplyr::select(cellcode, hash.ID) %>% 
  # Add count column and normalize to total hashtag size
  mutate(count = 1) %>% 
  group_by(hash.ID)%>%
  mutate(normalized.count = (count/sum(count))*1000 ) %>% 
  ungroup() %>% 
  # collapse genotypes
  mutate(genotype = fct_collapse(hash.ID,WT = c("HTO1", "HTO2", "HTO3"), 
                                 KO = c("HTO4", "HTO5", "HTO6"))) -> sample.counts

sample.counts
```

Next, join with the pseudotime data, split the normalized counts to separate columns, arrange by pseudotime and calculate the rolling sums.

```{r prep_data2}
sample.counts %>% 
  left_join(pseudotimes, by = "cellcode") %>% 
  pivot_wider(names_from = genotype,values_from =  normalized.count, values_fill = 0) %>% 
  arrange(desc(pseudotime))  %>%
  # Get rolling sums
  mutate(WT_sums = roll_sum(x = WT, n = 60, by = 1, align = "center", fill = NA), 
         KO_sums = roll_sum(x = KO, n = 60, by = 1, align = "center" , fill = NA)) -> rolling.sums

rolling.sums
```

And back to long data, and plot.

```{r plot_rolling, fig.width=7}
rolling.sums %>% 
  pivot_longer(cols = ends_with("sums"), names_to = "QPCTL_status", values_to = "cell_count_in_window") %>% 
    ggplot(aes(x = pseudotime, y = cell_count_in_window, color = QPCTL_status))+
      scale_x_reverse()+
      geom_line(size = 1)

ggsave(filename =  here("Figs", "immune", "moving_window_Pseudotime.pdf"), width = 6,height = 3, useDingbats = F )
```

## Analysis of pseudotime terminals

Another way of analysis would be to specifically look at the two terminals of pseudotime and compare those. First check cell counts from the samples

```{r get_terminals}
sample.counts %>% 
  # Add pseudotime data
  left_join(pseudotimes, by = "cellcode") %>% 
  ## subset on the right and left most terminals
  filter(dense_rank(pseudotime) <= 120 | dense_rank(dplyr::desc(pseudotime)) <= 120) %>% 
  mutate(terminal = case_when(pseudotime < 75 ~ "right",
                              pseudotime > 75 ~ "left")) %>% 
  group_by( genotype,terminal, hash.ID) %>% 
  summarise(sum.count = sum(normalized.count)) -> terminal.counts

terminal.counts
```

And plot. 

```{r plot_term_counts}
ggplot(terminal.counts, aes(x = genotype, y = sum.count))+
  geom_jitter(width = 0.2, size = 2)+
  facet_wrap(~terminal, scales = "free_y")+
  stat_summary(fun.data=mean_se,  geom="pointrange", color="blue", cex = 0.4 )

ggsave(here("Figs", "immune", "Pseudotime_terminals_bySide.pdf"), width = 3, height = 4, scale = .75, useDingbats = F)
```


And plot expression of genes of interest

```{r get_gene_dat}
GAM.data %>% 
  dplyr::filter(dense_rank(pseudotime) <= 120 | dense_rank(dplyr::desc(pseudotime)) <= 120) %>% 
  mutate(terminal = case_when(pseudotime < 75 ~ "right",
                              pseudotime > 75 ~ "left")) -> for.plot
```

```{r plot_term_expression}
for.plot %>% 
  dplyr::filter(gene %in% c("Ly6c2", "Mrc1", "Lyz1", "Lamp1", "Ms4a4c","Eps8", "Ctsd", "H2-Aa")) %>% 
  mutate(polarization = fct_recode(.$terminal, Mono = "left", Macro = "right")) %>% 
    ggplot( aes(x = polarization, y = logcount))+
      geom_violin(aes(fill = polarization))+
      geom_jitter(color = "black", size = 0.1, width = 0.25)+
      facet_wrap(~fct_relevel(gene, c("Ly6c2", "Mrc1", "Lyz1", "Lamp1", "Ms4a4c","Eps8", "Ctsd", "H2-Aa")), nrow = 2)+
      theme_classic()+
      theme(strip.background = element_blank())


ggsave(here("Figs", "immune", "Vln_plot_Phenotypes_terminals.pdf"), width = 5, height = 2.5, useDingbats = F)

```

